lib/model/narrow.dart

```diff
@@ -19,6 +19,18 @@ sealed class Narrow {
 /// A non-interleaved narrow, completely specifying a place to send a message.
 sealed class SendableNarrow extends Narrow {
   MessageDestination get destination;
+
+  static SendableNarrow ofMessage(Message message, {required int selfUserId}) {
+    switch (message) {
+      case StreamMessage(:var streamId, :var subject):
+        return TopicNarrow(streamId, subject);
+      case DmMessage(:var displayRecipient):
+        return DmNarrow(
+          allRecipientIds: displayRecipient.map((r) => r.id).toList(),
+          selfUserId: selfUserId,
```

Comment:
Makes sense.

> These can be found among the call sites of the `TopicNarrow` and `DmNarrow` constructors.

I see that a call site of the `TopicNarrow` constructor makes the `allRecipientIds` list with `.toList(growable: false)`. If `DmNarrow` is meant to be immutable, then that seems helpful: you don't want that list to get anything added to it.

…I guess you also don't want any of its elements to be replaced or removed, either. Would a `List.unmodifiable` be better? What about just an `Iterable`?

Reply:
> Would a list made with `List.unmodifiable` be better?

Yeah, that'd be pretty reasonable.

> What if `DmNarrow.allRecipientIds` were an `Iterable` instead of a `List`?

I think `DmNarrow` wants to ask that the `allRecipientIds` it's passed is a `List`, because the expectation is that it's all materialized and there's not any work that needs to be done to iterate through them.

It could provide the `allRecipientIds` property as an `Iterable` instead of a `List`, but I'm not sure that's useful enough for the small complication it adds.